### PR TITLE
Feature/oed interface

### PR DESCRIPTION
**Pull request by sstruzik**
Reworking of ods tools to merge all functionality into one consistent interface

### CODE REVIEW

**Comment by sambles** [comment]
Note: add pandas as package requirement

**Comment by sambles** [comment]
Add package requirements for:
```
pandas 
chardet 
```
```

**Comment by mtazzari** [comment]
@sambles in your review, can you please check the GH actions workflows? You're more familiar with them. Thanks!

**Comment by sambles** [comment]
> @sambles in your review, can you please check the GH actions workflows? You're more familiar with them. Thanks!

@mtazzari I merge an update on the ODS testing side already https://github.com/OasisLMF/ODS_OpenExposureData/pull/95 , just looking at the oasislmf at the moment

File: docker/extract_spec.py [hunk 1]
```diff
@@ -1,14 +1,60 @@
...
+
+
+@cli.command('csv')
+@click.option('--source-excel-path', required=True, default=None, help='Path to MS excel sheet')
+@click.option('--output-csv-path', default='OpenExposureData_Spec.csv', help='Path to write csv file')
+@click.option('--excel-sheet-name', default="OED Input Fields", help='Sheet label to extract')
 def extract_spec_to_csv(source_excel_path, output_csv_path, excel_sheet_name):
+    """
+    convert an Excel sheet to a csv file
+    Args:
```
Comment (mtazzari) [inline docker/extract_spec.py:52]:
```suggestion

    Args:
```

**Review by sambles — COMMENTED** [review]
Issue running the following: 

ods_tools (https://github.com/OasisLMF/ODS_OpenExposureData/pull/108)
oasislmf (https://github.com/OasisLMF/OasisLMF/pull/1171) 
OasisPlatform (latest) 
PiWind (develop)

[platform_ods_interface_run-error.txt](https://github.com/OasisLMF/ODS_OpenExposureData/files/10196202/platform_ods_interface_run-error.txt)

**Review by mtazzari — CHANGES_REQUESTED** [review]
Very general comment is why we don't we have `ods_tools` in a separate repository instead of having it inside ODS_OpenExposureData.

A second comment is that I would rename `src` into `ods_tools`. `src` is C/C++ style legacy, and very generic. It'd be just neater to have it `ods_tools`.
 
Specific comments:
- I haven't tested/evaluated the design patterns you adopted here, so I cannot advise whether it would have been better to take different approach for the use cases. I'm not familiar enough with the typical usage, so I trust this was the best way to implement the functionality.
- docstrings: can you please do an overhaul of the docstrings throughout the code and amend them so that they follow the format we've agreed upon? See e.g. my suggestions in `docker/extract_spec.py` and in src/oed/exposure.py`. I've found docstrings that are about right, but many of them are either missing a description (eg src/oed/schema OedSchema.is_valid_vaue), or the input types, etc. I have commented on many but not all the docstrings that need to be fixed, please do check them all.
- please check throughout the code some typos: occupency -> occupancy, unkown -> unknown. I've fixed those that I could see but it's better to do a full text search with an IDE and catch all the occurrences. 
- there's a filename renaming: `src/tests/__init__.py` -> `src/data/.gitkeep`   is it correct?
- I find it confusing to have a mixture of "OED", "oed", "Oed" throughout code and documentation. Can we reduce the representations? Eg use OED consistently throughout code and documentation? and perhaps `oed` only for command line arguments?

To speed things up on your side, I've added many suggestions to the review, which you can accept or reject with just a click, then when you've passed all the suggestions, you can pull the code locally and do the other changes.

File: docker/extract_spec.py [hunk 2]
```diff
@@ -1,14 +1,60 @@
...
+        source_excel_path: path to the Excel ods_schema
+        output_csv_path: output path
+        excel_sheet_name: name of the sheet to extract
+
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply docker/extract_spec.py]: done

File: docker/extract_spec.py [hunk 3]
```diff
@@ -22,5 +68,257 @@ def extract_spec_to_csv(source_excel_path, output_csv_path, excel_sheet_name):
...
+    """
+    get information on occupancy codes
+    Args:
+        occupency_df: occupancy Dataframe
+
+    Returns:
+        dict of information on occupancies
+    """
+    return (
+        occupency_df
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply docker/extract_spec.py]: done

File: src/README.md [hunk 4]
```diff
@@ -1,19 +1,213 @@
...
+Once the config is done you can create your OedExposure Object
+and access the Dataframe representation of the different sources.
+Data Type in the DataFrame will correspond to the type
+
+```python
+import ods_tools
+oed_exposure = ods_tools.oed.OedExposure(**config)
+location = oed_exposure.location.dataframe
+account = oed_exposure.account.dataframe
+```
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply src/README.md]: after discussion dataframe vs df was equivalent so we keep the existing solution dataframe

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    command.add_argument('--ri-scope', help='Path to the ri_scope file', default=None)
+    command.add_argument('--oed-dir', help='Path to the oed directory containing stardard named oed files', default=None)
+    command.add_argument('--config-json', help='Path to the config_json file', default=None)
+    command.add_argument('--validation-config', help='Path to the validation_config file', default=None)
+
+
+main_parser = argparse.ArgumentParser()
+
+oed_exposure_creation = """
+there is several option to specify the exposure data,
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply src/main.py]: done

File: src/main.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ - passing the path to the directory where the exposure is stored using oed_dir
+ - passing an oed config json
+ - passing the path to each oed source.
+ if both are specified the specific path will overwrite the eventual source found looking in oed_dir.
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply src/main.py]: should be better now

File: docker/extract_spec.py [hunk 7]
```diff
@@ -22,5 +68,257 @@ def extract_spec_to_csv(source_excel_path, output_csv_path, excel_sheet_name):
...
+    with open(output_json_path, 'w') as fp:
+        json.dump(ods_schema, fp, indent='    ')
+
+
+def get_ods_input_fields(ods_fields_df: pd.DataFrame):
+    """
+    Read OED Input Fields information, provide detailed  field information per file type as a dict
+
+    Args:
+        ods_fields_df: OED Input Fields sheet as a Dataframe
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply docker/extract_spec.py]: done

File: docker/extract_spec.py [hunk 8]
```diff
@@ -22,5 +68,257 @@ def extract_spec_to_csv(source_excel_path, output_csv_path, excel_sheet_name):
...
+
+
+def get_ods_perils(perils_df: pd.DataFrame):
+    """
+    information on ods perils
+    Args:
+        perils_df: Peril Values DataFrame
+
+    Returns:
+        dict with info on perils and mapping between peril and list of sub perils
```
[the remark this answers is not in the document]
Comment (sstruzik) [reply docker/extract_spec.py]: done

**Review by sambles — APPROVED** [review]

**Pull request merged**